File: dataforge-core/src/main/java/hep/dataforge/workspace/tasks/PipeTask.java
package hep.dataforge.workspace.tasks;

import hep.dataforge.actions.OneToOneAction;
import hep.dataforge.context.Context;
import hep.dataforge.data.DataNode;
import hep.dataforge.meta.Laminate;
import hep.dataforge.meta.Meta;

public abstract class PipeTask<T, R> extends AbstractTask<R> {
    private final String name;
    private final Class<T> inputType;
    private final Class<R> outputType;

    private final PipeAction action = new PipeAction();

    protected PipeTask(String name, Class<T> inputType, Class<R> outputType) {
        this.name = name;
        this.inputType = inputType;
        this.outputType = outputType;
    }


    @Override
    protected final DataNode<R> run(TaskModel model, DataNode<?> data) {
        return action.run(model.getContext(), data.checked(inputType), model.getMeta());
    }

    @Override
    protected abstract void buildModel(TaskModel.Builder model, Meta meta);

    protected abstract R result(Context context, String name, T input, Laminate meta);

    @Override
    public final String getName() {
        return name;
    }

    private class PipeAction extends OneToOneAction<T, R> {

        @Override
        public String getName() {
            return PipeTask.this.getName();
        }

        @Override
        public Class<T> getInputType() {
            return inputType;
        }

        @Override
        public Class<R> getOutputType() {
            return outputType;
        }

        @Override
        protected R execute(Context context, String name, T input, Laminate inputMeta) {
            return result(context, name, input, inputMeta);
        }
    }
}

File: dataforge-core/src/main/kotlin/hep/dataforge/workspace/tasks/PipeTask.kt
package hep.dataforge.workspace.tasks

import hep.dataforge.actions.OneToOneAction
import hep.dataforge.context.Context
import hep.dataforge.data.DataNode
import hep.dataforge.meta.Laminate
import hep.dataforge.meta.Meta

abstract class PipeTask<T : Any, R : Any> protected constructor(override val name: String, private val inputType: Class<T>, outputType: Class<R>) : AbstractTask<R>(outputType) {

    private val action = PipeAction()


    override fun run(model: TaskModel, data: DataNode<Any>): DataNode<R> {
        return action.run(model.context, data.checked(inputType), model.meta)
    }

    abstract override fun buildModel(model: TaskModel.Builder, meta: Meta)

    protected abstract fun result(context: Context, name: String, input: T, meta: Laminate): R

    private inner class PipeAction : OneToOneAction<T, R>() {

        override val name: String = this@PipeTask.name

        override fun getInputType(): Class<T> {
            return this@PipeTask.inputType
        }

        override fun getOutputType(): Class<R> {
            return this@PipeTask.type
        }

        override fun execute(context: Context, name: String, input: T, inputMeta: Laminate): R {
            return result(context, name, input, inputMeta)
        }
    }
}
